File: src/StaticDeque.zig
const std = @import("std");

pub fn StaticDeque(comptime T: type, comptime capacity: usize) type {
    return struct {
        const Self = @This();
        values: [capacity]T = undefined,
        capacity: usize = capacity,
        begin: usize = 0,
        end: usize = 0,

        pub fn pushBack(deque: *Self, value: T) !void {
            if (deque.size() == capacity) {
                return error.OutOfMemory;
            }
            deque.values[deque.end] = value;
            deque.end = (deque.end + 1) % capacity;
        }

        pub fn popBack(deque: *Self) ?T {
            if (deque.size() == 0) {
                return null;
            }
            deque.end = (deque.end + capacity - 1) % capacity;
            return deque.values[deque.end];
        }

        pub fn pushFront(deque: *Self, value: T) !void {
            if (deque.size() == capacity) {
                return error.OutOfMemory;
            }
            deque.begin = (deque.begin + capacity - 1) % capacity;
            deque.values[deque.begin] = value;
        }

        pub fn popFront(deque: *Self) ?T {
            if (deque.size() == 0) {
                return null;
            }
            const value = deque.values[deque.begin];
            deque.begin = (deque.begin + 1) % capacity;
            return value;
        }

        pub fn size(deque: *Self) usize {
            return (deque.end + capacity - deque.begin) % capacity;
        }

        pub fn isEmpty(deque: *Self) bool {
            return deque.size() == 0;
        }

        pub fn isFull(deque: *Self) bool {
            return deque.size() == capacity;
        }

        pub fn toSlice(deque: *Self) []T {
            if (deque.size() == 0) {
                return deque.values[0..0];
            }

            var temp: [capacity]T = undefined;

            var index: usize = 0;
            var current = deque.begin;
            while (index < deque.size()) : (index += 1) {
                temp[index] = deque.values[current];
                current = (current + 1) % capacity;
            }

            @memcpy(deque.values[0..deque.size()], temp[0..deque.size()]);

            deque.begin = 0;
            deque.end = deque.size();

            return deque.values[0..deque.size()];
        }
    };
}

File: ts/src/staticDeque.ts
export class StaticDeque {
  private readonly values: Int32Array;
  private begin = 0;
  private end = 0;
  private length = 0;

  constructor(private readonly capacity: number) {
    if (!Number.isInteger(capacity) || capacity <= 0) {
      throw new RangeError("StaticDeque: capacity must be positive integer");
    }
    this.values = new Int32Array(capacity);
  }

  pushBack(value: number): void {
    if (this.length === this.capacity) {
      throw new RangeError("StaticDeque overflow");
    }
    this.values[this.end] = value;
    this.end = (this.end + 1) % this.capacity;
    this.length += 1;
  }

  popBack(): number | null {
    if (this.length === 0) return null;
    this.end = (this.end + this.capacity - 1) % this.capacity;
    const value = this.values[this.end];
    this.length -= 1;
    return value ?? null;
  }

  pushFront(value: number): void {
    if (this.length === this.capacity) {
      throw new RangeError("StaticDeque overflow");
    }
    this.begin = (this.begin + this.capacity - 1) % this.capacity;
    this.values[this.begin] = value;
    this.length += 1;
  }

  popFront(): number | null {
    if (this.length === 0) return null;
    const value = this.values[this.begin];
    this.begin = (this.begin + 1) % this.capacity;
    this.length -= 1;
    return value ?? null;
  }

  size(): number {
    return this.length;
  }

  isEmpty(): boolean {
    return this.length === 0;
  }

  isFull(): boolean {
    return this.length === this.capacity;
  }

  toArray(): Int32Array {
    const result = new Int32Array(this.length);
    for (let i = 0; i < this.length; i += 1) {
      const idx = (this.begin + i) % this.capacity;
      result[i] = this.values[idx]!;
    }
    this.begin = 0;
    this.end = this.length;
    this.values.set(result);
    return result;
  }
}
